src/service/resources-impl.js

```diff
@@ -578,6 +590,70 @@ export class Resources {
     }
   }
 
+  /**
+   * Builds resources during prerender in the document order. Ignores children
+   * of a resource that does not allow prerendering.
+   * @private
+   */
+  dfsPrerender_() {
+    if (!this.dfsPrerenderTreeWalker_) {
+      this.dfsPrerenderTreeWalker_ = this.win.document.createTreeWalker(
+          this.ampdoc.getBody(), NodeFilter.SHOW_ELEMENT,
+          {acceptNode: this.dfsPrerenderTreeWalkerFilter_.bind(this)}, false);
+    }
+
+    while (this.buildAttemptsCount_ <= MAX_BUILDS_IN_PRERENDER &&
+        this.dfsPrerenderTreeWalker_.nextNode()) {
+      const node = this.dfsPrerenderTreeWalker_.currentNode;
+
+      // Node isn't upgraded, set the walker head to the last upgraded node, or
+      // to the walker root, so the next pass checks its status again.
+      if (!(node.isUpgraded && node.isUpgraded())) {
+        this.dfsPrerenderTreeWalker_.previousNode();
+
+        if (node === this.dfsPrerenderTreeWalker_.currentNode) {
+          this.dfsPrerenderTreeWalker_.currentNode =
+              this.dfsPrerenderTreeWalker_.root;
+        }
+
+        return;
+      }
+
+      const r = Resource.forElementOptional(node);
+      if (!r.isBuilt() && !r.isBuilding()) {
+        this.buildResourceUnsafe_(r, false);
+      }
+    }
+  }
+
+  /**
+   * DFS TreeWalker filter, finding DOM trees containing resources that allow
+   * prerendering.
+   * @param {!Element} element
+   * @return {!NodeFilter}
+   * @private
+   */
+  dfsPrerenderTreeWalkerFilter_(element) {
```

Comment:
Using a filter is slower than handling this in your while loop. That's because it has to call from C++ (the walker) to the JS (the filter) to accept back to C++ (walker stops walking) back to JS (walker returns node to while loop).

Just use the while loop.